torch/csrc/jit/fusion_compiler.cpp

```diff
@@ -87,11 +93,302 @@ struct TensorInfo {
+      RaiseKey(&key);
+      counter = ComputeSingleRound(counter, key);
+      RaiseKey(&key);
+      counter = ComputeSingleRound(counter, key);
+      RaiseKey(&key);
+      counter = ComputeSingleRound(counter, key);
+      RaiseKey(&key);
+      counter = ComputeSingleRound(counter, key);
+      RaiseKey(&key);
+      counter = ComputeSingleRound(counter, key);
+      RaiseKey(&key);
+      counter = ComputeSingleRound(counter, key);
+      RaiseKey(&key);
+      counter = ComputeSingleRound(counter, key);
+      RaiseKey(&key);
+      counter = ComputeSingleRound(counter, key);
+
+      SkipOne();
+
+      return counter;
+    }
+
+   private:
+    // We use the same constants as recommended by the original paper.
+    static const uint32 kPhiloxW32A = 0x9E3779B9;
+    static const uint32 kPhiloxW32B = 0xBB67AE85;
+    static const uint32 kPhiloxM4x32A = 0xD2511F53;
+    static const uint32 kPhiloxM4x32B = 0xCD9E8D57;
+
+    // Helper function to skip the next sample of 128-bits in the current stream.
+    PHILOX_DEVICE_INLINE void SkipOne() {
+      if (++counter_[0] == 0) {
+        if (++counter_[1] == 0) {
+          if (++counter_[2] == 0) {
+            ++counter_[3];
+          }
+        }
+      }
+    }
+
+    // Helper function to return the lower and higher 32-bits from two 32-bit
+    // integer multiplications.
+    PHILOX_DEVICE_INLINE
+    static void MultiplyHighLow(uint32 a, uint32 b, uint32* result_low,
+                                uint32* result_high) {
+  #ifndef __CUDA_ARCH__
+      const uint64 product = static_cast<uint64>(a) * b;
+      *result_low = static_cast<uint32>(product);
+      *result_high = static_cast<uint32>(product >> 32);
+  #else
+      *result_low = a * b;
+      *result_high = __umulhi(a, b);
+  #endif
+    }
+
+    // Helper function for a single round of the underlying Philox algorithm.
+    PHILOX_DEVICE_INLINE static ResultType ComputeSingleRound(
+        const ResultType& counter, const Key& key) {
+      uint32 lo0;
+      uint32 hi0;
+      MultiplyHighLow(kPhiloxM4x32A, counter[0], &lo0, &hi0);
+
+      uint32 lo1;
+      uint32 hi1;
+      MultiplyHighLow(kPhiloxM4x32B, counter[2], &lo1, &hi1);
+
+      ResultType result;
+      result[0] = hi1 ^ counter[1] ^ key[0];
+      result[1] = lo1;
+      result[2] = hi0 ^ counter[3] ^ key[1];
+      result[3] = lo0;
+      return result;
+    }
+
+    PHILOX_DEVICE_INLINE void RaiseKey(Key* key) {
+      (*key)[0] += kPhiloxW32A;
+      (*key)[1] += kPhiloxW32B;
+    }
+
+   private:
+    ResultType counter_;
+    Key key_;
+  };
+
+  // A Wrapper for Philox to get one random number at a time
+  class PhiloxWrapper {
+  public:
+    PHILOX_DEVICE_INLINE
+    explicit PhiloxWrapper(uint64 seed_lo, uint64 seed_hi, uint64 offset) {
+      philox = PhiloxRandom(seed_lo, seed_hi);
+      philox.Skip(offset / 4);
+      i = 0;
+    }
+    PHILOX_DEVICE_INLINE float operator()() {
+      if(i == 0) buf = philox();
+      uint32 ret = buf[i];
```

Comment:
This is a dynamic index into a per-thread array, which will cause that array to be stored in local memory, and kill your performance by a factor of 2 or more.  Replace with the following:
```
    uint32 ret;
    switch(i)
    {
      default:
        ret = buf[0];
        break;
    case 1:
        ret = buf[1];
        break;
    case 2:
        ret = buf[2];
        break;
    case 3:
        ret = buf[3];
        break;
    }
```